src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,462 @@
+        if let Err(e) = register_timer(event_loop, Tick::ReportStatistic, REPORT_STATISTIC_INTERVAL) {
+            error!("register report statistic err: {:?}", e);
+        };
+    }
+
+    fn on_report_staticstic_tick(&self, event_loop: &mut EventLoop<Self>) {
+        info!("all running cmd count = {}", self.cmd_ctxs.len());
+
+        self.register_report_tick(event_loop);
+    }
+
+    fn on_received_new_cmd(&mut self, cmd: Command) {
+        let cid = self.gen_id();
+        let lock = self.gen_lock(&cmd);
+        let ctx = RunningCtx::new(cid, cmd, lock);
+        self.save_cmd_context(cid, ctx);
+
+        if self.acquire_lock(cid) {
+            self.get_snapshot(cid);
         }
-        Command::Prewrite { ctx, mutations, primary, start_ts, callback } => {
-            callback(match store.prewrite(ctx, mutations, primary, start_ts) {
-                Ok(mut results) => {
-                    Ok(results.drain(..).map(|x| x.map_err(::storage::Error::from)).collect())
+    }
+
+    fn acquire_lock(&mut self, cid: u64) -> bool {
+        let ctx = &mut self.cmd_ctxs.get_mut(&cid).unwrap();
+        self.latches.acquire(&mut ctx.lock, cid)
+    }
+
+    fn get_snapshot(&mut self, cid: u64) {
+        let ch = self.schedch.clone();
+        let cb = box move |snapshot: EngineResult<Box<Snapshot>>| {
+            if let Err(e) = ch.send(Msg::SnapshotFinish{ cid: cid, snapshot: snapshot, }) {
+                error!("send SnapshotFinish failed cmd id {}, err {:?}", cid, e);
+            }
+        };
+
+        if let Err(e) = self.engine.async_snapshot(self.extract_context(cid), cb) {
+            self.finish_with_err(cid, Error::from(e));
+        }
+    }
+
+    fn on_snapshot_finished(&mut self, cid: u64, snapshot: EngineResult<Box<Snapshot>>) {
+        debug!("receive snapshot finish msg for cid = {}", cid);
+
+        match snapshot {
+            Ok(snapshot) => {
+                let res = self.process_cmd_with_snapshot(cid, snapshot.as_ref());
+
+                let finished = match res {
+                    Ok(()) => {
+                        let ctx = self.cmd_ctxs.get(&cid).unwrap();
+                        ctx.cmd.readonly()
+                    }
+                    Err(e) => {
+                        self.process_failed_cmd(cid, Error::from(e));
+                        true
+                    }
+                };
+                if finished {
+                    self.finish_cmd(cid);
                 }
-                Err(e) => Err(e.into()),
-            });
+            }
+            Err(e) => { self.finish_with_err(cid, Error::from(e)); }
         }
-        Command::Commit { ctx, keys, lock_ts, commit_ts, callback } => {
-            callback(store.commit(ctx, keys, lock_ts, commit_ts)
-                .map_err(::storage::Error::from));
+    }
+
+    fn on_write_finished(&mut self, cid: u64, pr: ProcessResult, result: EngineResult<()>) {
```

Comment:
Async procedure is hard to debug or trace, please add some logs when enter&leave each process stage.